common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -471,9 +488,10 @@ public void onData(String streamId, ByteBuffer buf) {
         public void onComplete(String streamId) {
           if (isStaleBlockOrTooLate) {
             // Throw an exception here so the block data is drained from channel and server
-            // responds RpcFailure to the client.
-            throw new RuntimeException(String.format("Block %s %s", streamId,
-              ErrorHandler.BlockPushErrorHandler.TOO_LATE_OR_STALE_BLOCK_PUSH_MESSAGE_SUFFIX));
+            // responds the error code to the client.
+            throw new BlockPushNonFatalFailure(
+              new PushBlockNonFatalErrorCode(ErrorCode.TOO_LATE_OR_STALE_BLOCK_PUSH.id())
```

Comment:
Shall we use the individual error codes for "too late push" and "stale push" separately?

Reply:
This is doable with some minor changes on the `RemoteBlockPushResolver` side to tell apart the two errors.

Reply:
I am not quite sure whether it is possible to separate out `too late` and `stale` pushes separately.

Reply:
May be it is possible with the introduction of `StaleBlockPushException`.

Reply:
Is there a need for `StaleBlockPushException` any more? The new  `BlockPushNonFatalFailure` is a `RuntimeException` and if there are separate error codes for `too late` and `stale push` then having a `StaleBlockPushException` seems unnecessary. Can it be removed?

Reply:
This is used with in `RemoteBlockPushResolver`, we need it to reject older `shuffleMergeId` block push requests. See this https://github.com/apache/spark/blob/d8169493b662acac31d3fc5e6c5051917428c974/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java#L399 This is required since we cannot return null in the `compute` as that would remove the `shuffleId` Also see this comment - https://github.com/apache/spark/pull/33034#discussion_r676098927

Reply:
I saw how it was being used. My point is that this is not needed any more. We can use `BlockPushNoneFatalFailure` instead of `StaleBlockPushException`. 

Also,  catching the `StaleBlockPushException` is not really required.  I am pointing to the below lines. If `getOrCreateAppShufflePartitionInfo` throws a `RuntimeException`, then `receiveBlockDataAsStream` would just fail. In case of stale push, it's not really needed that a `StreamCallback` is returned.
```
    try {
      partitionInfoBeforeCheck = getOrCreateAppShufflePartitionInfo(appShuffleInfo, msg.shuffleId,
        msg.shuffleMergeId, msg.reduceId);
    } catch(StaleBlockPushException sbp) {
      // Set partitionInfoBeforeCheck to null so that stale block push gets handled.
      partitionInfoBeforeCheck = null;
    }
 ```
 So, this could just simplified with 
 ```
     AppShufflePartitionInfo partitionInfoBeforeCheck = getOrCreateAppShufflePartitionInfo(appShuffleInfo, msg.shuffleId,
        msg.shuffleMergeId, msg.reduceId);
 ```
 If the execution goes beyond this, then that means the push was not stale and then it is just too late.

Reply:
@otterc you are right that we can just replace `StaleBlockPushException` with `BlockPushNonFatalFailure`. Catching it and later throwing another is not necessary.

Reply:
@Victsm that's a good point. But we can still replace the usage of `StaleBlockPushException` which is created with a message that uses `String.format` with `BlockPushNonFatalFailure`.  We can still catch `BlockPushNonFatalFailure` and based on the `ErrorCode` enum differentiate between stale and too late.

Also, for stale block pushes that are for prior application attempts, isn't it  closing the channel better? For clients that are pushing blocks from prior application attempt, we just want them to stop pushing all together correct.

Reply:
For stale push from prior app attempt, it currently indeed throws an IllegalArgumentException inside `receiveBlockDataAsStream` which would lead to server closing channels.

Reply:
> For stale push from prior app attempt, it currently indeed throws an IllegalArgumentException inside receiveBlockDataAsStream which would lead to server closing channels.

I see. That's right. I missed that stale push here is used only in reference to the executors which are for the same app attempts but older merged id. 

But anyways, I still think we should replace `StaleBlockPushException` usage with `BlockPushNonFatalFailure`.